types/mempool/priority.go

```diff
@@ -0,0 +1,234 @@
+	nonce := sig.Sequence
+	tk := txKey{nonce: nonce, priority: ctx.Priority(), sender: sender}
+
+	senderIndex, ok := mp.senderIndices[sender]
+	if !ok {
+		senderIndex = huandu.New(huandu.LessThanFunc(func(a, b any) int {
+			return huandu.Uint64.Compare(b.(txKey).nonce, a.(txKey).nonce)
+		}))
+		// initialize sender index if not found
+		mp.senderIndices[sender] = senderIndex
+	}
+
+	// Since senderIndex is scored by nonce, a changed priority will overwrite the existing txKey.
+	senderIndex.Set(tk, tx)
+
+	// Since mp.priorityIndex is scored by priority, then sender, then nonce, a changed priority will create a new key,
+	// so we must remove the old key and re-insert it to avoid having the same tx with different priorityIndex indexed
+	// twice in the mempool.  This O(log n) remove operation is rare and only happens when a tx's priority changes.
+	sk := txKey{nonce: nonce, sender: sender}
+	if oldScore, txExists := mp.scores[sk]; txExists {
+		mp.priorityIndex.Remove(txKey{nonce: nonce, priority: oldScore, sender: sender})
+	}
+	mp.scores[sk] = ctx.Priority()
+	mp.priorityIndex.Set(tk, tx)
+
+	return nil
+}
+
+// Select returns a set of transactions from the mempool, ordered by priority and sender-nonce in O(n) time.
+// The passed in list of transactions are ignored.  This is a readonly operation, the mempool is not modified.
+// maxBytes is the maximum number of bytes of transactions to return.
+func (mp *priorityMempool) Select(_ [][]byte, maxBytes int64) ([]Tx, error) {
+	var selectedTxs []Tx
+	var txBytes int64
+	senderCursors := make(map[string]*huandu.Element)
+
+	// start with the highest priority sender
+	priorityNode := mp.priorityIndex.Front()
+	for priorityNode != nil {
+		priorityKey := priorityNode.Key().(txKey)
+		nextHighestPriority, nextPriorityNode := nextPriority(priorityNode)
+		sender := priorityKey.sender
+		senderTx := mp.fetchSenderCursor(senderCursors, sender)
+
+		// iterate through the sender's transactions in nonce order
+		for senderTx != nil {
+			mempoolTx := senderTx.Value.(Tx)
+			if mp.onRead != nil {
+				mp.onRead(mempoolTx)
+			}
+
+			k := senderTx.Key().(txKey)
+
+			// break if we've reached a transaction with a priority lower than the next highest priority in the pool
+			if k.priority < nextHighestPriority {
+				break
+			}
+
+			// otherwise, select the transaction and continue iteration
+			selectedTxs = append(selectedTxs, mempoolTx)
+			if txBytes += mempoolTx.Size(); txBytes >= maxBytes {
+				return selectedTxs, nil
+			}
+
+			senderTx = senderTx.Next()
+			senderCursors[sender] = senderTx
+		}
+
+		priorityNode = nextPriorityNode
+	}
+
+	return selectedTxs, nil
+}
+
+func (mp *priorityMempool) fetchSenderCursor(senderCursors map[string]*huandu.Element, sender string) *huandu.Element {
+	senderTx, ok := senderCursors[sender]
+	if !ok {
+		senderTx = mp.senderIndices[sender].Front()
+	}
+	return senderTx
+}
+
+func nextPriority(priorityNode *huandu.Element) (int64, *huandu.Element) {
```

Comment:
nit: even though both of these are private functions/methods, they are pretty important so I think they need pretty good godocs on what they do and their purpose.